Proofreading-OV-Runtime (#11658)
openvinotoolkit/openvino: docs/OV_Runtime_UG/preprocessing_details.md

> # Preprocessing API - Details {#openvino_docs_OV_UG_Preprocessing_Details}

Reviewer: ```suggestion
Pre-processing API - Details {#openvino_docs_OV_UG_Preprocessing_Details}
```

> 

Reviewer: ```suggestion

The purpose of this article is to present details on preprocessing API, such as its capabilities and post-processing.

```

> ## Preprocessing Capabilities

Reviewer: ```suggestion
## Pre-processing Capabilities
```

> 

Reviewer: ```suggestion

Below is a full list of pre-processing API capabilities:

```

### Addressing Particular Input/Output

> If the model has only one input, then simple `ov::preprocess::PrePostProcessor::input()` will get a reference to preprocessing builder for this input (a tensor, the steps, a model):

Reviewer: ```suggestion
If the model has only one input, then simple `ov::preprocess::PrePostProcessor::input()` will get a reference to pre-processing builder for this input (a tensor, the steps, a model):
```

@endsphinxtabset

C++ references:
  * `ov::preprocess::InputTensorInfo`
  * `ov::preprocess::OutputTensorInfo`
  * `ov::preprocess::PrePostProcessor`

> ### Supported Preprocessing Operations

Reviewer: ```suggestion
### Supported Pre-processing Operations
```

@endsphinxtabset

C++ references:
* `ov::preprocess::PreProcessSteps::mean()`
* `ov::preprocess::PreProcessSteps::scale()`

#### Converting Precision

In Computer Vision, image is represented by an array of unsigned 8-but integer values (for each color), but model accepts floating point tensors.

> To integrate precision conversion into an execution graph as a preprocessing step, simply do the following:

Reviewer: ```suggestion
To integrate precision conversion into an execution graph as a pre-processing step, simply do the following:
```

@endsphinxtabset

It performs the same transpose. However, the approach where source and destination layout are used can be easier to read and understand.

C++ references:
  * `ov::preprocess::PreProcessSteps::convert_layout()`
  * `ov::preprocess::InputTensorInfo::set_layout()`
  * `ov::preprocess::InputModelInfo::set_layout()`
  * `ov::Layout`

#### Resizing Image

> Resizing of an image is a typical preprocessing step for computer vision tasks. With preprocessing API, this step can also be integrated into execution graph and performed on a target device.

Reviewer: ```suggestion
Resizing of an image is a typical pre-processing step for computer vision tasks. With pre-processing API, this step can also be integrated into execution graph and performed on a target device.
```

@endsphinxtabset

C++ references:
* `ov::preprocess::PreProcessSteps::resize()`
* `ov::preprocess::ResizeAlgorithm`

#### Color Conversion

> Typical use case is to reverse color channels from `RGB` to `BGR` and vice versa. To do this, specify source color format in `tensor` section and perform `convert_color` preprocessing operation. In the example below, a `BGR` image needs to be converted to `RGB` as required for the model input.

Reviewer: ```suggestion
Typical use case is to reverse color channels from `RGB` to `BGR` and vice versa. To do this, specify source color format in `tensor` section and perform `convert_color` pre-processing operation. In the example below, a `BGR` image needs to be converted to `RGB` as required for the model input.
```

@endsphinxtabset

#### Color Conversion - NV12/I420

> Preprocessing also supports YUV-family source color formats, i.e. NV12 and I420.

Reviewer: ```suggestion
Pre-processing also supports YUV-family source color formats, i.e. NV12 and I420.
```

> In advanced cases, such YUV images can be split into separate planes, e.g. for NV12 images Y-component may come from one source and UV-component from another one. Concatenating such components in user's application manually is not a perfect solution from performance and device utilization perspectives. However, there is a way to use Preprocessing API. For such cases there are `NV12_TWO_PLANES` and `I420_THREE_PLANES` source color formats, which will split the original `input` into 2 or 3 inputs.

Reviewer: ```suggestion
In advanced cases, such YUV images can be split into separate planes, e.g., for NV12 images Y-component may come from one source and UV-component from another one. Concatenating such components in user's application manually is not a perfect solution from performance and device utilization perspectives. However, there is a way to use Pre-processing API. For such cases there are `NV12_TWO_PLANES` and `I420_THREE_PLANES` source color formats, which will split the original `input` into 2 or 3 inputs.
```

@endsphinxtabset

In this example, the original `input` is split to `input/y` and `input/uv` inputs. You can fill `input/y` from one source, and `input/uv` from another source. Color conversion to `RGB` will be performed, using these sources. It is more efficient as there will be no additional copies of NV12 buffers.

C++ references:
* `ov::preprocess::ColorFormat`
* `ov::preprocess::PreProcessSteps::convert_color`

### Custom Operations

> Preprocessing API also allows adding custom preprocessing steps into execution graph. Custom step is a function, which accepts current `input` node and returns new node after adding preprocessing step.

Reviewer: ```suggestion
Pre-processing API also allows adding custom preprocessing steps into execution graph. Custom step is a function, which accepts current `input` node and returns new node after adding preprocessing step.
```

@endsphinxtabset

C++ references:
* `ov::preprocess::PreProcessSteps::custom()`
* [Available Operations Sets](../ops/opset.md)

> ## Postprocessing

Reviewer: ```suggestion
## Post-processing
```

> Postprocessing steps can be added to model outputs. As for preprocessing, these steps will be also integrated into a graph and executed on a selected device.

Reviewer: ```suggestion
Post-processing steps can be added to model outputs. As for pre-processing, these steps will be also integrated into a graph and executed on a selected device.
```

> Preprocessing uses the following flow:  **User tensor** -> **Steps** -> **Model input**.

Reviewer: ```suggestion
Pre-processing uses the following flow: **User tensor** -> **Steps** -> **Model input**.
```

> Postprocessing uses the reverse:  **Model output** -> **Steps** -> **User tensor**.

Reviewer: ```suggestion
Post-processing uses the reverse: **Model output** -> **Steps** -> **User tensor**.
```

Compared to preprocessing, there are not so many operations needed for post-processing stage. Currently, only the following postprocessing operations are supported:
 - Convert a [layout](./layout_overview.md).
 - Convert an element type.
 - Customize operations.

> Usage of these operations is similar to Preprocessing. Below is an example:

Reviewer: ```suggestion
Usage of these operations is similar to pre-processing. Below is an example:
```
